### ISSUE TITLE
Rule proposal: `prefer-smaller-scope` (like `prefer-const`, but across scopes)

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#1524]
Context:

- https://github.com/eslint/eslint/issues/14900
- https://github.com/eslint/eslint/issues/5284

This is basically what `prefer-const` does:

```js
function foo () {
	  let a; // ERROR: Use `const` instead (prefer-const)
	  a = 1;
	  return a;
}
```

Except that is should work across scopes

## Fail

```js
function foo () {
	let a;
	while (Math.random() > 0.5) {
		a = get(); // ERROR: Use `const` instead (prefer-smaller-scope)
		console.log(a);
	}
}
```

## Pass

```js
function foo () {
	while (Math.random() > 0.5) {
		const a = get(); // ✅
		console.log(a);
	}
}
```

```js
function foo () {
	let a;
	while (Math.random() > 0.5) {
		a = get(a); // ✅ `a` is used before assignment 
		console.log(a);
	}
}
```

```js
function foo () {
	let a;
	while (Math.random() > 0.5) {
		a = get();
	}
	console.log(a); // ✅ `a` is actually used in the outer scope
}
```

**Comment by sindresorhus** [comment]
@fisker What's your verdict on this?

**Comment by fisker** [comment]
I think this can be useful, but I have two cases I don't want the variable inside the function: 1. A constant, that's a big array, object, map, set, or a expensive call. 2. A object to cache things, this won't work if it moved into the function, I don't know how to detect that.

**Comment by fisker** [comment]
Only check `let` not `const`?

**Comment by brettz9** [comment]
FWIW: https://github.com/eslint/eslint/issues/12992
